cpp/include/cudf/detail/null_mask.cuh

```diff
@@ -199,251 +217,287 @@ __global__ void subtract_set_bits_range_boundaries_kernel(bitmask_type const* bi
                                               0,
                                               stream.value()>>>(
-    bitmask, num_ranges, d_first_indices.begin(), d_last_indices.begin(), d_null_counts.begin());
+    bitmask, num_ranges, first_bit_indices_begin, last_bit_indices_begin, d_bit_counts.begin());
+
+  if (count_bits == count_bits_policy::UNSET_BITS) {
+    // Convert from set bits counts to unset bits by subtracting the number of
+    // set bits from the length of the segment.
+    auto segments_begin =
+      thrust::make_zip_iterator(first_bit_indices_begin, last_bit_indices_begin);
+    auto segments_size = thrust::transform_iterator(segments_begin, [] __device__(auto segment) {
+      auto const begin = thrust::get<0>(segment);
+      auto const end   = thrust::get<1>(segment);
+      return end - begin;
+    });
+    thrust::transform(rmm::exec_policy(stream),
+                      segments_size,
+                      segments_size + num_ranges,
+                      d_bit_counts.data(),
+                      d_bit_counts.data(),
+                      [] __device__(auto segment_size, auto segment_bit_count) {
+                        return segment_size - segment_bit_count;
+                      });
+  }
 
   CHECK_CUDA(stream.value());
+  return d_bit_counts;
+}
 
-  std::vector<size_type> ret(num_ranges);
-  CUDA_TRY(cudaMemcpyAsync(ret.data(),
-                           d_null_counts.data(),
-                           num_ranges * sizeof(size_type),
-                           cudaMemcpyDeviceToHost,
-                           stream.value()));
+/**
+ * @brief Given two iterators, validate that the iterators represent valid ranges of
+ * indices and return the number of ranges.
+ *
+ * @throws cudf::logic_error if `std::distance(indices_begin, indices_end) % 2 != 0`
+ * @throws cudf::logic_error if `indices_begin[2*i] < 0 or indices_begin[2*i] >
+ * indices_begin[(2*i)+1]`
+ *
+ * @param indices_begin An iterator representing the beginning of the ranges of indices
+ * @param indices_end An iterator representing the end of the ranges of indices
+ *
+ * @return The number of segments specified by the input iterators.
+ */
+template <typename IndexIterator>
+size_type validate_segmented_indices(IndexIterator indices_begin, IndexIterator indices_end)
+{
+  auto const num_indices = static_cast<size_type>(std::distance(indices_begin, indices_end));
+  CUDF_EXPECTS(num_indices % 2 == 0, "Array of indices needs to have an even number of elements.");
+  size_type const num_segments = num_indices / 2;
+  for (size_type i = 0; i < num_segments; i++) {
+    auto begin = indices_begin[2 * i];
+    auto end   = indices_begin[2 * i + 1];
+    CUDF_EXPECTS(begin >= 0, "Starting index cannot be negative.");
+    CUDF_EXPECTS(end >= begin, "End index cannot be smaller than the starting index.");
+  }
+  return num_segments;
+}
 
-  stream.synchronize();  // now ret is valid.
+struct index_alternator {
```

Comment:
There are at least three places in this PR where I would benefit from a strided iterator (input validation in `validate_segmented_indices`, constructing first/last index iterators from the iterator of segmented indices in `segmented_count_bits`, and constructing a vector of segment lengths in `segmented_valid_count`). I would like to work on https://github.com/NVIDIA/thrust/issues/912 in the near future. I might introduce a strided iterator into `cudf/detail/iterator.cuh` to test the design before working on a Thrust implementation.